### PR TITLE
Add WGSL support for slang-test

### PR DESCRIPTION
**Pull request by aleino-nv**
- Add Tint as a downstream compiler
- Add some plumbing for `slang-test` to target WebGPU and WGSL using the newly added RHI support
- Disable several asserting `slang-test` tests in `tests/compute`

### CODE REVIEW

**Comment by aleino-nv** [comment]
@csyonghe Should I use the `breaking` label instead? I'm adding some stuff to the middle of an enum in slang.h.
Maybe I should just move that value to the end?

**Comment by aleino-nv** [comment]
There are some things that seem a bit strange about this: I have added "WGSL SPIR-V" and "WGSL SPIR-V assembly" targets in a bunch of places just to be able to express that I want Tint to produce the SPIR-V.

@csyonghe 
Maybe there are other mechanisms to achieve this that don't involve adding these strange "redundant" targets?

**Comment by csyonghe** [comment]
To answer your questions: all enums need to be added at the end to avoid binary breakage. Defining wasm-spirv and wasm-spirv-asm as separate targets is fine.

**Comment by csyonghe** [comment]
Instead of adding `DISABLED_TEST` lines to individual .slang files, please add the test names to `tests/expected-failure-github.txt` so we can use the list as a centralized place to track which tests are being disabled.

**Comment by aleino-nv** [comment]
> Instead of adding `DISABLED_TEST` lines to individual .slang files, please add the test names to `tests/expected-failure-github.txt` so we can use the list as a centralized place to track which tests are being disabled.

I can do that, but these tests are _asserting_ which seems to interrupt the test run a lot. Disabling gets rid of that problem.

Do you still think I should just mark these as test failures?

**Comment by aleino-nv** [comment]
> > Instead of adding `DISABLED_TEST` lines to individual .slang files, please add the test names to `tests/expected-failure-github.txt` so we can use the list as a centralized place to track which tests are being disabled.
> 
> I can do that, but these tests are _asserting_ which seems to interrupt the test run a lot. Disabling gets rid of that problem.
> 
> Do you still think I should just mark these as test failures?

As discussed on Slack I will keep these disabled and then also add to the expected-failure list.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [inline source/compiler-core/slang-tint-compiler.cpp:109]: Use String instead of std::string.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [inline source/compiler-core/slang-tint-compiler.cpp:82]: No const

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [inline source/compiler-core/slang-tint-compiler.cpp:83]: This leaks. Use RefPtr<TintDownstreamCompiler>

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [inline source/compiler-core/slang-tint-compiler.cpp:92]: No const.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [inline source/compiler-core/slang-tint-compiler.cpp:118]: Leaking memory on tint_compileResult.

Comment (csyonghe) [reply source/compiler-core/slang-tint-compiler.cpp:118]: Use SLANG_DEFER to ensure things are freed:

auto rs = m_compile(&result);
SLANG_DEFER(m_free(&result));
if (rs != 0) ...

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [inline source/compiler-core/slang-tint-compiler.cpp:117]: int resultCode =

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [inline source/compiler-core/slang-tint-compiler.cpp:121]: No var{} syntax for initialization.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [inline source/compiler-core/slang-tint-compiler.cpp:141]: No new line after return.

File: source/core/slang-render-api-util.cpp
Comment (csyonghe) [reply source/core/slang-render-api-util.cpp:272]: This isn't addressed.

File: source/slang/slang.cpp
Comment (csyonghe) [reply source/slang/slang.cpp:266]: This isnt addressed.

**Review by csyonghe — CHANGES_REQUESTED** [review]

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:118]: Do you mean m_compile fails? A common convention (which is followed in this case) is to make sure that no resources are actually returned if the call fails.

    extern "C"
    SLANG_SHARED_LIB_EXPORT
    int tint_compile(tint_CompileRequest* request, tint_CompileResult* result)
    {
        ///...
        uint8_t *const resultBuffer {static_cast<uint8_t*>(::malloc(resultBufferSize))};
        if(resultBuffer == nullptr)
            return 1;
        ::memcpy(resultBuffer, spirvCode.data(), resultBufferSize);
    
        result->buffer = resultBuffer;
        result->bufferSize = resultBufferSize;
        return 0;
    }

Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:118]: (But of course I may as well add a DEFER anyway.)

**Comment by csyonghe** [comment]
On coding style:
No ); on its own line.
No ( on its own line.
No const unless necessary (required by c++ standard).
No T x{} initialization. Should be T c= {}
No return on its own line.

Long function signature should either be
 ```
Blah longFunc(
     All parameters);
```
Or
```
Blah longFunc(
   Param1
   ParamN);
```

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [reply source/compiler-core/slang-tint-compiler.cpp:118]: How do we report back diagnostics returned by tint?

File: source/core/slang-render-api-util.cpp
Comment (aleino-nv) [reply source/core/slang-render-api-util.cpp:272]: Dawn compiles the WGSL we give it into something that it can use to render/compute. On Windows by default that is Directx12. In my WebGPU test application, Dawn wouldn't compile pipelines unless it could find dxcompiler.

File: source/slang/slang.cpp
Comment (aleino-nv) [reply source/slang/slang.cpp:266]: As discussed in Slack, I will add a comment.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:109]: Done

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:83]: Done.

The class doesn't implement `RefObject`, which seems required for `RefPtr`. I switched to using `ComPtr<IDownstreamCompiler>` like the analogous code in slang-metal-compiler.cpp.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:92]: Done

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:118]: Done. I added an error field to the slang-tint result struct.


https://github.com/shader-slang/dawn/blob/slang-tint/src/slang-tint/slang-tint.cpp#L62

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:117]: Done

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:121]: Done.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:141]: Done.

File: source/compiler-core/slang-glslang-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-glslang-compiler.cpp:304]: Done, I think.

File: external/slang-tint-headers/slang-tint.h
Comment (aleino-nv) [reply external/slang-tint-headers/slang-tint.h:8]: Done

File: source/compiler-core/slang-glslang-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-glslang-compiler.cpp:306]: Done, I think.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:78]: I'm confused.
Here is one example from https://github.com/shader-slang/slang/blob/master/docs/design/coding-conventions.md?plain=1#L178

    float gVar = bigFunc(
        0,
        1.0f,
        data);

Do you want it to look like this (A)?

        SLANG_RETURN_ON_FAIL(
            DownstreamCompilerUtil::loadSharedLibrary(path,
                loader, nullptr, "slang-tint", library));
                
Or like this (B)?

        SLANG_RETURN_ON_FAIL(DownstreamCompilerUtil::loadSharedLibrary(
            path, loader, nullptr, "slang-tint", library));



Doing B for now.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:128]: My "too long" is currently set to 90 based on this:

"There is no hard limit on line length, but if you are going past 80-100 columns quite often, maybe think about breaking lines."


I changed it to this instead:

        ComPtr<IArtifact> resultArtifact = ArtifactUtil::createArtifactForCompileTarget(
            SlangCompileTarget::SLANG_WGSL_SPIRV);

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:139]: Done

File: source/core/slang-render-api-util.cpp
Comment (aleino-nv) [reply source/core/slang-render-api-util.cpp:270]: Done

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [reply source/compiler-core/slang-tint-compiler.cpp]: The codebase follows B.

File: source/compiler-core/slang-tint-compiler.cpp
Comment (csyonghe) [reply source/compiler-core/slang-tint-compiler.cpp]: This is good.

**Review by csyonghe — DISMISSED** [review]

File: source/compiler-core/slang-tint-compiler.cpp
Comment (aleino-nv) [reply source/compiler-core/slang-tint-compiler.cpp:116]: Done.

**Comment by csyonghe** [comment]
Reading the ci log it doesn't seem wgpu is reported as supported and no tests are run for wgpu.

**Review by csyonghe — COMMENTED** [review]
Please make sure all tests are run on the wgpu backend similar to how tests are run on metal. Right now no tests are run on the ci even with up to date slang-rhi.

We have logic in slang-test that synthesizes new tests for all platforms not explicitly specified by the test. For example, we will run tests on metal even if the test doesn't specify it explicitly. We want the same logic for wgpu.

**Comment by aleino-nv** [comment]
> Please make sure all tests are run on the wgpu backend similar to how tests are run on metal. Right now no tests are run on the ci even with up to date slang-rhi.
> 
> We have logic in slang-test that synthesizes new tests for all platforms not explicitly specified by the test. For example, we will run tests on metal even if the test doesn't specify it explicitly. We want the same logic for wgpu.

I filed a separate task for CI integration, which I'm working on next https://github.com/shader-slang/slang/issues/5233.

Is there a reason for this to hold up the current patch?
As soon as this patch goes in, at least people can easily run tests locally.

**Comment by skallweitNV** [comment]
> Is there a reason for this to hold up the current patch?
> As soon as this patch goes in, at least people can easily run tests locally.

Agreed, would make my life easier.

File: include/slang-gfx.h
Comment (skallweitNV) [inline include/slang-gfx.h:95]: this should probably be removed, there is no WebGPU support for gfx

**Comment by csyonghe** [comment]
Merge this PR now to unblock @westlicht.

**Pull request merged**
